core/src/main/java/org/apache/iceberg/Variants.java

```diff
@@ -0,0 +1,207 @@
+    BOOLEAN,
+    EXACT_NUMERIC,
+    FLOAT,
+    DOUBLE,
+    DATE,
+    TIMESTAMPTZ,
+    TIMESTAMPNTZ,
+    BINARY,
+    STRING,
+    ARRAY,
+    OBJECT
+  }
+
+  public enum PhysicalType {
+    NULL(LogicalType.NULL, Void.class),
+    BOOLEAN_TRUE(LogicalType.BOOLEAN, Boolean.class),
+    BOOLEAN_FALSE(LogicalType.BOOLEAN, Boolean.class),
+    INT8(LogicalType.EXACT_NUMERIC, Integer.class),
+    INT16(LogicalType.EXACT_NUMERIC, Integer.class),
+    INT32(LogicalType.EXACT_NUMERIC, Integer.class),
+    INT64(LogicalType.EXACT_NUMERIC, Long.class),
+    DOUBLE(LogicalType.DOUBLE, Double.class),
+    DECIMAL4(LogicalType.EXACT_NUMERIC, BigDecimal.class),
+    DECIMAL8(LogicalType.EXACT_NUMERIC, BigDecimal.class),
+    DECIMAL16(LogicalType.EXACT_NUMERIC, BigDecimal.class),
+    DATE(LogicalType.DATE, Integer.class),
+    TIMESTAMPTZ(LogicalType.TIMESTAMPTZ, Long.class),
+    TIMESTAMPNTZ(LogicalType.TIMESTAMPNTZ, Long.class),
+    FLOAT(LogicalType.FLOAT, Float.class),
+    BINARY(LogicalType.BINARY, ByteBuffer.class),
+    STRING(LogicalType.STRING, String.class),
+    ARRAY(LogicalType.ARRAY, List.class),
+    OBJECT(LogicalType.OBJECT, Map.class);
+
+    private final LogicalType logicalType;
+    private final Class<?> javaClass;
+
+    PhysicalType(LogicalType logicalType, Class<?> javaClass) {
+      this.logicalType = logicalType;
+      this.javaClass = javaClass;
+    }
+
+    LogicalType toLogicalType() {
+      return logicalType;
+    }
+
+    public Class<?> javaClass() {
+      return javaClass;
+    }
+
+    public static PhysicalType from(int primitiveType) {
+      switch (primitiveType) {
+        case Primitives.TYPE_NULL:
+          return NULL;
+        case Primitives.TYPE_TRUE:
+          return BOOLEAN_TRUE;
+        case Primitives.TYPE_FALSE:
+          return BOOLEAN_FALSE;
+        case Primitives.TYPE_INT8:
+          return INT8;
+        case Primitives.TYPE_INT16:
+          return INT16;
+        case Primitives.TYPE_INT32:
+          return INT32;
+        case Primitives.TYPE_INT64:
+          return INT64;
+        case Primitives.TYPE_DATE:
+          return DATE;
+        case Primitives.TYPE_TIMESTAMPTZ:
+          return TIMESTAMPTZ;
+        case Primitives.TYPE_TIMESTAMPNTZ:
+          return TIMESTAMPNTZ;
+        case Primitives.TYPE_FLOAT:
+          return FLOAT;
+        case Primitives.TYPE_DOUBLE:
+          return DOUBLE;
+        case Primitives.TYPE_DECIMAL4:
+          return DECIMAL4;
+        case Primitives.TYPE_DECIMAL8:
+          return DECIMAL8;
+        case Primitives.TYPE_DECIMAL16:
+          return DECIMAL16;
+        case Primitives.TYPE_BINARY:
+          return BINARY;
+        case Primitives.TYPE_STRING:
+          return STRING;
+      }
+
+      throw new UnsupportedOperationException("Unknown primitive physical type: " + primitiveType);
+    }
+  }
+
+  public interface Serialized {
+    ByteBuffer buffer();
+  }
+
+  public interface Metadata extends Serialized {
```

Comment:
I think it makes sense to have VariantPrimitive, VariantArray, VariantObject classes which implements VariantAccessor (or similar name). We probably shouldn't expose Metadata or Value directly since those are more internal representation of a Variant.

Similarly `buffer()` may not need to be exposed as well if possible?

Reply:
As part of cleaning up the API, I've made this package-private. The methods that matter are `writeTo` and `sizeInBytes` on `VariantValue`.